api/python/quilt3/formats.py

```diff
@@ -1024,6 +1027,62 @@ def deserialize(self, bytes_obj, meta=None, ext=None, **format_opts):
 ParquetFormatHandler().register()  # latest is preferred
 
 
+# noinspection PyPackageRequirements
+class AnnDataFormatHandler(BaseFormatHandler):
+    """Format for AnnData <--> .h5ad
+
+    Format Opts:
+        The following options may be used anywhere format opts are accepted,
+        or directly in metadata under `{'format': {'opts': {...: ...}}}`.
+
+        compression('gzip', 'lzf', None):  applies during serialization only.
+    """
+    name = 'h5ad'
+    handled_extensions = ['h5ad']
+    opts = ('compression',)
+    defaults = dict(
+        compression='lzf',
+    )
+
+    def handles_type(self, typ: type) -> bool:
+        # don't load module unless we actually have to use it.
+        if 'anndata' not in sys.modules:
+            return False
+        import anndata as ad
+        self.handled_types.add(ad.AnnData)
+        return super().handles_type(typ)
+
+    def serialize(self, obj, meta=None, ext=None, **format_opts):
+        opts = self.get_opts(meta, format_opts)
+        opts_with_defaults = copy.deepcopy(self.defaults)
+        opts_with_defaults.update(opts)
+        try:
+            buf = io.BytesIO()
+            obj.write(buf, **opts_with_defaults)
+            data = buf.getvalue()
+        except TypeError:
+            # Old AnnData version don’t support writing to buffers,
+            # see https://github.com/scverse/anndata/pull/800
+            with NamedTemporaryFile() as f:
+                f.close()  # can’t re-open a file on Windows
+                obj.write(f.name, **opts_with_defaults)
+                data = Path(f.name).read_bytes()
```

Comment:
I think it doesn't work the way you expected, file is deleted only once the first time `close()` is called. For example:

```python
with tempfile.NamedTemporaryFile() as f:
    f.close()
    with open(f.name, 'x'):
        pass
with open(f.name):  # This is expected to fail, but works.
    pass

Reply:
fixed!